pkg/mutation/mutators/assignimage/imageparser_test.go

```diff
@@ -0,0 +1,372 @@
+				tag:    ":latest",
+			},
+		},
+		{
+			name:     "only location",
+			imageRef: "some-image/hello",
+			want: image{
+				domain: "",
+				path:   "some-image/hello",
+				tag:    "",
+			},
+		},
+		{
+			name:     "no slash in location",
+			imageRef: "some-image:tag123",
+			want: image{
+				domain: "",
+				path:   "some-image",
+				tag:    ":tag123",
+			},
+		},
+		{
+			name:     "just location",
+			imageRef: "alpine",
+			want: image{
+				domain: "",
+				path:   "alpine",
+				tag:    "",
+			},
+		},
+		{
+			name:     "leading underscore",
+			imageRef: "_/alpine",
+			want: image{
+				domain: "",
+				path:   "_/alpine",
+				tag:    "",
+			},
+		},
+		{
+			name:     "leading underscore with tag",
+			imageRef: "_/alpine:latest",
+			want: image{
+				domain: "",
+				path:   "_/alpine",
+				tag:    ":latest",
+			},
+		},
+		{
+			name:     "no domain, location has /",
+			imageRef: "library/busybox:v9",
+			want: image{
+				domain: "",
+				path:   "library/busybox",
+				tag:    ":v9",
+			},
+		},
+		{
+			name:     "dots in domain",
+			imageRef: "this.that.com/repo/alpine:1.23",
+			want: image{
+				domain: "this.that.com",
+				path:   "repo/alpine",
+				tag:    ":1.23",
+			},
+		},
+		{
+			name:     "port and dots in domain",
+			imageRef: "this.that.com:5000/repo/alpine:latest",
+			want: image{
+				domain: "this.that.com:5000",
+				path:   "repo/alpine",
+				tag:    ":latest",
+			},
+		},
+		{
+			name:     "localhost with port",
+			imageRef: "localhost:5000/repo/alpine:latest",
+			want: image{
+				domain: "localhost:5000",
+				path:   "repo/alpine",
+				tag:    ":latest",
+			},
+		},
+		{
+			name:     "dots in location",
+			imageRef: "x.y.z/gcr.io/repo:latest",
+			want: image{
+				domain: "x.y.z",
+				path:   "gcr.io/repo",
+				tag:    ":latest",
+			},
+		},
+		{
+			name:     "dot in domain",
+			imageRef: "gcr.io/repo:latest",
+			want: image{
+				domain: "gcr.io",
+				path:   "repo",
+				tag:    ":latest",
+			},
+		},
+		{
+			name:     "invalid ref still parses",
+			imageRef: "x.io/get/ready4.//this/not.good:404@yikes/bad/string",
+			want: image{
+				domain: "x.io",
+				path:   "get/ready4.//this/not.good",
+				tag:    ":404@yikes/bad/string",
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			got := newImage(tc.imageRef)
+			if got != tc.want {
+				t.Errorf("got: %v, want %v", got, tc.want)
+			}
+		})
+	}
+}
+
+func TestValidateImageParts(t *testing.T) {
+	tests := []struct {
+		name    string
+		domain  string
+		path    string
+		tag     string
+		wantErr bool
+	}{
+		{
+			name:   "all valid components",
+			domain: "my.register.io:5000",
+			path:   "lib/stuff/app",
+			tag:    ":latest",
+		},
+		{
+			name:    "no fields set returns err",
+			wantErr: true,
+		},
+		{
+			name:   "valid domain with port",
+			domain: "localhost:5000",
+		},
+		{
+			name:   "valid domain no port",
+			domain: "localhost",
```

Comment:
right now, I believe `foobar` would be a valid domain, we should test to make sure it raises an error

Reply:
regex updated to handle this; tests added